src/utils/daos_hdlr.c

```diff
@@ -1535,6 +1536,741 @@ cont_destroy_hdlr(struct cmd_args_s *ap)
+	int rc = dfs_release(dirp->dir);
+	if (rc) {
+		fprintf(stderr, "dfs_release failed (%d %s)",
+			rc, strerror(rc));
+		errno = rc;
+		rc = -1;
+	}
+	free(dirp);
+	return rc;
+}
+
+static int
+daos_file_closedir(daos_file_t *daos_file, DIR* dirp)
+{
+	int rc = 0;
+	if (daos_file->type == POSIX) {
+		rc = closedir(dirp);
+	} else if (daos_file->type == DAOS) {
+		rc = daos_closedir(dirp);
+	} else {
+		fprintf(stderr, "File type not known, type=%d", daos_file->type);
+	}
+	return rc;
+}
+
+static int
+daos_close(daos_file_t *daos_file, const char* file)
+{
+	int rc = dfs_release(daos_file->obj);
+	if (rc) {
+		fprintf(stderr, "dfs_close %s failed (%d %s)",
+			file, rc, strerror(rc));
+		errno = rc;
+		rc = -1;
+	}
+	return rc;
+}
+
+static int
+daos_file_close(daos_file_t *daos_file, const char *file)
+{
+	int rc = 0;
+	if (daos_file->type == POSIX) {
+		rc = close(daos_file->fd);
+		if (rc == 0)
+			daos_file->fd = -1;
+	} else if (daos_file->type == DAOS) {
+		rc = daos_close(daos_file, file);
+		if (rc == 0)
+			daos_file->obj = NULL;
+	} else {
+		fprintf(stderr, "File type not known, file=%s, type=%d",
+			file, daos_file->type);
+	}
+	return rc;
+}
+
+static void
+fs_copy(daos_file_t *daos_src_file,
+		daos_file_t *daos_dst_file,
+		const char* dir_name,
+		const char* fs_dst_prefix)
+{
+	DIR *d;
+	int rc;
+
+	/* begin by opening source directory */
+	d = daos_file_opendir(daos_src_file, dir_name);
+
+	 /* check it was opened. */
+	if (!d) {
+		fprintf (stderr, "Cannot open directory '%s': %s\n",
+			dir_name, strerror (errno));
+	}
+
+	while (1) {
+		struct dirent *entry;
+		const char *d_name;
+
+		/* walk source directory */
+		entry = daos_file_readdir(daos_src_file, d);
+		if (!entry) {
+			/* There are no more entries in this directory, so break
+			 * out of the while loop. */
+			break;
+		}
+
+		d_name = entry->d_name;
+		char filename[MAX_FILENAME];
+		char dst_filename[MAX_FILENAME];
+		snprintf(filename, MAX_FILENAME, "%s/%s", dir_name, d_name);
+		struct stat st;
+		if (daos_src_file->type == DAOS && daos_dst_file->type == POSIX)
+			snprintf(dst_filename, MAX_FILENAME, "%s/%s", fs_dst_prefix, filename);
+
+		/* stat the source file */
+		rc = daos_file_stat(daos_src_file, filename, &st);
+		if (rc)
+			fprintf(stderr, "Cannot stat path %s, %s\n",
+				d_name, strerror(errno));
+
+		if (S_ISREG(st.st_mode)) {
+			if (daos_src_file->type == DAOS && daos_dst_file->type == POSIX) {
+				daos_file_open(daos_src_file, filename, O_CREAT | O_RDWR, st.st_mode);
+				daos_file_open(daos_dst_file, dst_filename, O_CREAT | O_RDWR, st.st_mode);
+			} else {
+				daos_file_open(daos_src_file, filename, O_CREAT | O_RDWR, st.st_mode);
+				daos_file_open(daos_dst_file, filename, O_CREAT | O_RDWR, st.st_mode);
+			}
+
+			/* read from source file, then write to destination file */
+			uint64_t total_bytes = st.st_size;
+			uint64_t left_to_read = 0;
+			uint64_t buf_size = 64*1024*1024;
+			void* buf = malloc(buf_size * sizeof(char));
```

Comment:
```suggestion
			D_ALLOC(buf, buf_size * sizeof(char));
				if (buf == NULL)
					return ENOMEM;
```
I think we can use `D_ALLOC`, and also we need to check for `ENOMEM`